kmsjwt.go

```diff
@@ -2,149 +2,75 @@ package kmsjwt
+	"github.com/aws/aws-sdk-go-v2/service/kms/types"
+	"github.com/golang-jwt/jwt/v4"
 	"github.com/pkg/errors"
 )
 
-const kmsAlgorighm = "KMS"
-
-// ErrKmsVerification is an error shown when KMS token verification fails.
-var ErrKmsVerification = errors.New("kms: verification error")
-
-// ErrInvalidKey indicates taht the key is invalid.
-var ErrInvalidKey = errors.New("key is invalid")
-
+// KMSJWT is a JWT signing method implementation using RSA512 with the private
+// key stored in AWS KMS. The public key is retrieved from KMS on
+// initialization.
 type KMSJWT struct {
-	kmsiface.KMSAPI
-
-	algorithm         string
-	cache             *cache.Cache
-	kmsKeyID          string
-	withCache         bool
-	defaultExpiration time.Duration
-	cleanupInterval   time.Duration
-	signingAlgorithm  string
+	api       KMS
+	keyID     string
+	publicKey *rsa.PublicKey
 }
 
 // New provides a KMS-based implementation of JWT signing method.
-func New(client kmsiface.KMSAPI, kmsKeyID string, opts ...Option) *KMSJWT {
-	ret := &KMSJWT{
-		KMSAPI:            client,
-		algorithm:         kmsAlgorighm,
-		kmsKeyID:          kmsKeyID,
-		withCache:         true,
-		defaultExpiration: time.Hour,
-		cleanupInterval:   time.Minute,
-		signingAlgorithm:  kms.SigningAlgorithmSpecRsassaPssSha512,
+func New(ctx context.Context, api KMS, keyID string) (*KMSJWT, error) {
+	out, err := api.GetPublicKey(ctx, &kms.GetPublicKeyInput{KeyId: aws.String(keyID)})
+	if err != nil {
+		return nil, errors.Wrap(err, "could not retrieve public key")
 	}
 
-	for _, opt := range opts {
-		opt(ret)
+	publicKey, err := jwt.ParseRSAPublicKeyFromPEM(out.PublicKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not parse public key")
 	}
 
-	if ret.withCache {
-		ret.cache = cache.New(ret.defaultExpiration, ret.cleanupInterval)
-	}
+	return &KMSJWT{api, keyID, publicKey}, nil
+}
 
-	return ret
+// NewWithPublicKey provides a KMS-based implementation of JWT signing method
+// with a pre-loaded public key.
+func NewWithPublicKey(api KMS, keyID string, publicKey *rsa.PublicKey) *KMSJWT {
+	return &KMSJWT{api, keyID, publicKey}
 }
 
 func (k *KMSJWT) Alg() string {
-	return k.algorithm
+	return jwt.SigningMethodRS512.Alg()
 }
 
 func (k *KMSJWT) Sign(signingString string, key interface{}) (string, error) {
 	ctx, ok := key.(context.Context)
 	if !ok {
-		return "", errors.New("key is not a context")
-	}
-
-	out, err := k.SignWithContext(ctx, &kms.SignInput{
-		KeyId:            aws.String(k.kmsKeyID),
-		Message:          checksum(signingString),
-		MessageType:      aws.String("DIGEST"),
-		SigningAlgorithm: aws.String(k.signingAlgorithm),
-	})
-
-	if err != nil && errors.Is(err, context.Canceled) {
-		return "", err
-	} else if err != nil {
-		return "", errors.Wrap(err, "key is invalid")
+		return "", jwt.ErrInvalidKeyType
 	}
 
-	if k.cache != nil {
-		k.cache.SetDefault(signingString, out.Signature)
-	}
+	hash := jwt.SigningMethodPS512.Hash.New()
```

Comment:
Shouldn't it be
```suggestion
	hash := jwt.SigningMethodRS512.Hash.New()
```
?

Could we make signing method a field that is defined only once, so that we are sure we don't get a one-letter typo?

Reply:
Great catch and suggestion. What threw me off guard in this case is that the hashing function is identical for both of these signing methods.

Anyway, fix and suggestion both implemented, PTAL 🙇🏻

Reply:
Also if that's indeed an error, could we add a test that will do Sign + Verify so that we are sure we are using the same algorithm here?

Reply:
Will do. I make the expected signature a const.

Reply:
🙇